### DISCUSSION TITLE
plan for typing of `structuredDiffForFile`

### DISCUSSION THREAD

**Discussion by berlysia** [danger/danger-js#1200] [General]
The return value of `structuredDiffForFile` is not well typed. (and I use this heavily)

https://github.com/danger/danger-js/blob/a7355a33d1f414b1d2e3140ca0c0062e711823ea/source/danger.d.ts#L809-L813

I did a little research and found that in general case, the `parse-diff` 's return object is used as is. see. https://github.com/danger/danger-js/issues/1089
The case of `BitBucketServer` is an exception and there seems to be some effort. see. https://github.com/danger/danger-js/issues/773

https://github.com/danger/danger-js/blob/a7355a33d1f414b1d2e3140ca0c0062e711823ea/source/platforms/bitbucket_server/BitBucketServerGit.ts#L77-L80

There is some variation of plan, as far as I can think:

A. Create `BitBucketServerGitDSL` or just like `GitDSL<BitBuckerServer>` , which have specialized type definition of `structredDiffForFile`.
B. Add type for general case. BitBucketServer users are required to cast type for their environment.
C. Add type for general case. Modify the implementation for BitBucketServer to fit the general case.
D. Add type as union of general case and specialized case. All user must choice to write type-narrowing code or to ignore types.
E. No change for code. Provide a comment to get well-typed `structuredDiffForFile`.
F. other

According to https://github.com/danger/danger-js/issues/773 , C is the most likely choice but there's still a difference.

How do you think about this?

**Comment by orta** [comment]
Yeah, I think C makes the most sense 👍🏻

**Reply by berlysia** [reply]
Great. I made https://github.com/danger/danger-js/pull/1201
Would you take a look?

**Comment by berlysia** [comment]
#1201
